File: basic_graph_logic.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

#

def walk(G, s, S=set()):
    '''Walking through a connected component of a graph represented using adjacency sets
    G: graph, s: start node, S: why needed?
    => P is return value. But it is hard to get the traveral path from s.
    '''
    P = dict()      # Predecessor
    Q = set()       # Queue => using set as Queue
    P[s] = None
    Q.add(s)

    while Q:
        u = Q.pop()
        for v in G[u].difference(P, S):
            Q.add(v)
            P[v] = u

    return P

def components(G):
    '''Return connected components
    '''
    comp = []
    seen = set()
    for u in G:
        if u in seen: continue

        c = walk(G, u)
        comp.append(c)
        seen.update(c)

    return comp


def main():
    ''' Code test
    '''
    G = {'a': set('bc'), 'b':set('ac'), 'c':set('ab'), 'd':set(), 'e':set('gh'), 'f':set('g'), 'g':set('ef'), 'h':set('e')}

    print(components(G))

if __name__ == '__main__':
    main()

File: basic_graph_logic.cpp
#include <iostream>
#include <map>
#include <set>
#include <queue>
#include <list>

using namespace std;

template<typename T>
void walk(const map<T, set<T>> &graph, T s, map<T, T> &pred )
{
    queue<T> q;

    pred[s] = -1;
    q.push(s);

    while (!q.empty()){
        T u = q.back();
        q.pop();

        for (const auto& s : graph.at(u)){
            for (const auto&  v : s){
                if (pred.count(v) > 0) continue;
                q.push(v);
                pred[v] = u;
            }
        }
    }
}

template<typename T>
void component(const map<T, set<T>> &graph, list<map<T, T>> &comp)
{
    set<T> seen;

    for(auto& v : graph){
        if (seen.count(v.first) > 0) continue;

        map<T, T> c;
        walk(graph, v.first, c);
        comp.push_back(c);
        for (auto &u : c){
            seen.insert(u.first);
        }
    }
}
int main(int argc, char const *argv[])
{
    // map initialization
    map<char, set<char>> graph = {{'a', {'b', 'c'}}, {'b', {'a', 'c'}}, {'c', {'a', 'b'}}, {'d', {}}, {'e', {'g', 'h'}}, {'f', {}}, {'g', {'e', 'f'}}, {'h', {'e'}}};

    list<map<char, char>> comp;
    component(graph, comp);

    // cout << comp << endl;
    return 0;
}
